src/object/cli_obj.c

```diff
@@ -3841,9 +3840,13 @@ obj_comp_cb(tse_task_t *task, void *data)
 	    DAOS_FAIL_CHECK(DAOS_DTX_NO_RETRY))
 		obj_auxi->io_retry = 0;
 
-	if (!obj_auxi->no_retry && (pm_stale || obj_auxi->io_retry))
-		obj_retry_cb(task, obj, obj_auxi, pm_stale,
-			     obj_auxi->map_ver_reply);
+	if (!obj_auxi->no_retry && (pm_stale || obj_auxi->io_retry)) {
+		rc = obj_retry_cb(task, obj, obj_auxi, pm_stale, obj_auxi->map_ver_reply);
+		if (rc) {
+			D_ERROR(DF_OID "retry io failed: %d\n", DP_OID(obj->cob_md.omd_id), rc);
+			obj_bulk_fini(obj_auxi);
```

Comment:
obj_bulk_fini() looks not enough for the cleanup, here may set "obj_auxi->io_retry = 0" so below cleanup code will be called?
BTW, looks quite some CI failures.

Reply:
The obj_bulk_fini() will be called in the subsequent section "if (!obj_auxi->io_retry) {". "io_retry" will be clear inside obj_retry_cb() if hit failure.

I cannot find at which point we may miss to call bulk_fini().

Reply:
you are right, since obj_auxi->io_retry will be cleared in fail case, seems need not add obj_bulk_fini() here. @wangdi1 ?

Reply:
I refreshed as nasf said, added an assertion here, and let below code do the cleanup.